### PR TITLE
add recommended Elixir and Erlang/OTP combinations to CI

### PR DESCRIPTION
**Pull request by adrianomitre**
### Main course

Make sure all recommended Elixir and Erlang/OTP combinations are run on CI

Quoting @josevalim:
> I would still advise folks to go with:
> * For each Elixir version, add a run with earliest supported Erlang/OTP
> * For the last Elixir version, also test the latest supported Erlang/OTP

Please refer to the original discussion on https://github.com/dashbitco/broadway/pull/188 for the full rationale.

### Side dish

This pull request also
- updates [actions/setup-elixir](https://github.com/actions/setup-elixir) and [actions/checkout](https://github.com/actions/checkout) to their latest versions
- restricts compilation with `--warnings-as-errors` to the latest Elixir version, on a separate CI job, following [suggestion](https://github.com/whatyouhide/stream_data/pull/142#discussion_r444396899) by @whatyouhide 
- add CI job to check formatting (`mix format --check-formatted`) on the latest Elixir version

### CODE REVIEW

**Comment by rrrene** [comment]
Hi there,

thx for putting this together! Unfortunately, I  do not think that I can merge these as they seem to yield no obvious benefit for this project.

Could you elaborate why you think these changes make sense for Credo's CI suite?

**Comment by adrianomitre** [comment]
Hi @rrrene,

For the rationale behing @josevalim recommendation, I would refer you to the [original discussion](https://github.com/dashbitco/broadway/pull/188).

In addition to the all that was pointed out [in the description above](https://github.com/rrrene/credo/pull/785#issue-440280999), I see the following benefits:
- testing with the latest official Elixir release 1.10.3
- testing with the latest official Erlang/OTP release 23.0.2
- more readable, less cyclomatically complex code by getting rid of the plethora of `ifs`

Finally, do you have any specific concern, so I may try to address it directly? Or do you have any suggestion on how to improve the PR?

**Comment by josevalim** [comment]
@adrianomitre please don't literally quote me on every discussion, because that will send me notifications. :D Maybe use my name literally as "José Valim". Thanks!

**Comment by adrianomitre** [comment]
Sorry, José Valim. :smirk:

**Comment by rrrene** [comment]
@adrianomitre It seems to me that this PR adds a lot of implicit complexity to get rid of a small bit of explicit complexity.

Testing fewer Elixir/Erlang combinations does have different implications for a tool like Credo than it has for a library like Broadway.

What I was asking was "What is the tangible benefit?" because it seemed to me that you simply quoted other maintainers regarding their decisions for their projects.

**Comment by adrianomitre** [comment]
> What I was asking was "What is the tangible benefit?" because it seemed to me that you simply quoted other maintainers regarding their decisions for their projects.

For one, it [reveals](https://github.com/rrrene/credo/pull/785/checks?check_run_id=817462216) that the test suite breaks on Elixir 1.5, which, according to [`mix.exs`](https://github.com/rrrene/credo/blob/master/mix.exs#L10) is supported.

File: .github/workflows/ci-workflow.yml [hunk 1]
```diff
@@ -12,30 +12,53 @@ jobs:
...
+          - otp: 20.3.8.26
+            elixir: 1.7.4
+          - otp: 20.3.8.26
+            elixir: 1.8.2
+          - otp: 20.3.8.26
+            elixir: 1.9.4
+          - otp: 21.3.8.16
+            elixir: 1.10.3
+          - otp: 23.0.2
+            elixir: 1.10.3
```
Comment (sgerrand) [inline .github/workflows/ci-workflow.yml:31]: In the spirit of "latest supported", perhaps this should be [`1.10.4`](https://github.com/elixir-lang/elixir/tree/v1.10.4)?

File: .github/workflows/ci-workflow.yml [hunk 2]
```diff
@@ -12,30 +12,53 @@ jobs:
...
     steps:
-      - uses: actions/checkout@v2
-        if: "!(matrix.otp == '20.3' && matrix.elixir == '1.10.0')"
-
-      - uses: actions/setup-elixir@v1
-        if: "!(matrix.otp == '20.3' && matrix.elixir == '1.10.0')"
+      - uses: actions/checkout@v2.3.1
+      - uses: actions/setup-elixir@v1.3
```
Comment (sgerrand) [inline .github/workflows/ci-workflow.yml:34]: 👏

File: .github/workflows/ci-workflow.yml [hunk 3]
```diff
@@ -12,30 +12,53 @@ jobs:
...
-        if: "!(matrix.otp == '20.3' && matrix.elixir == '1.10.0')"
+
+  compile_with_no_warnings:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.1
+      - uses: actions/setup-elixir@v1.3
+        with:
+          otp-version: 23.0.2
+          elixir-version: 1.10.3
```
Comment (sgerrand) [inline .github/workflows/ci-workflow.yml:49]: As per the previous comment on versions.

File: .github/workflows/ci-workflow.yml [hunk 4]
```diff
@@ -12,30 +12,53 @@ jobs:
...
+          ERLC_OPTS: "warnings_as_errors"
+
+  check_formatting:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.1
+      - uses: actions/setup-elixir@v1.3
+        with:
+          otp-version: 23.0.2
+          elixir-version: 1.10.3
```
Comment (sgerrand) [inline .github/workflows/ci-workflow.yml:63]: Also here.

File: test/smoke_test.sh [hunk 5]
```diff
@@ -4,8 +4,6 @@ set -e
 
 GENEREATE_CREDO_CHECK="lib/my_first_credo_check.ex"
 
-mix compile --force --warnings-as-errors
```
Comment (sgerrand) [inline test/smoke_test.sh:7]: 👏

File: .github/workflows/ci-workflow.yml (hunk 1 above)
Comment (adrianomitre) [reply .github/workflows/ci-workflow.yml]: Done in 4c3fa22d4ccf41697d06285c9e4cdfa75d6c855d

File: .github/workflows/ci-workflow.yml (hunk 4 above)
Comment (adrianomitre) [reply .github/workflows/ci-workflow.yml]: Done in 4c3fa22d4ccf41697d06285c9e4cdfa75d6c855d

File: .github/workflows/ci-workflow.yml (hunk 3 above)
Comment (adrianomitre) [reply .github/workflows/ci-workflow.yml]: Done in 4c3fa22d4ccf41697d06285c9e4cdfa75d6c855d

**Comment by rrrene** [comment]
> For one, it [reveals](https://github.com/rrrene/credo/pull/785/checks?check_run_id=817462216) that the test suite breaks on Elixir 1.5, which, according to [`mix.exs`](https://github.com/rrrene/credo/blob/master/mix.exs#L10), Credo supports.

I really, really do not want to discourage future contributions by yourself or other aspiring Elixir programmers.

But please reflect on what you are proposing here:

* you reduced the number of validated combinations in Credo's CI by a factor of three
* added two new build jobs because someone in another project said he prefers his build jobs this way in his project
* one of the new build jobs changes the semantics of Credo's build process
* the other one introduces a new fail condition
* and you break the smoke test shell script (probably without even noticing)

And all of this to "reveal" that the build - as is - would break for Elixir 1.5?

Take it from someone who has opened and reworked hundreds of PRs (and was rejected on probably just as many): A good first PR fixes a small thing and a good way to approach a possibly larger refactoring is to ask first.

As I said, I do not want to curb your enthusiasm, but I do not see the benefit of most of these changes and I do not like that they are presented in a single, convoluted PR citing other maintainers' preferences as argument to change fundamental principles of the build pipeline and workflow.

**Comment by adrianomitre** [comment]
Hi, @rrrene I have split the contributions of this pull request into a bunch of minimal pull requests with titles that make their purpose immediate. The overall theme is making sure that credo actually supports all Elixir and Erlang/OTP versions it claims to support.

**Pull request closed without merging**